### ISSUE TITLE
Bound DateEdit resets when date typed

### ISSUE THREAD

**Issue by vbray1979** [stsrki/Blazorise#1029]
**Describe the bug**
The date cannot be typed, as it keep resetting on typing only when bound to nullable DateTime

**To Reproduce**
Insert a bounded DateEdit component to nullable DateTime.
Start typing and date keep reseting

**Expected behavior**
Allow typing a date in a bounded DateEdit.

Remark: Works well id unbounded

**Comment by stsrki** [comment]
Can you provide an example?

**Comment by niem94** [comment]
I have this problem as well, I think. 

I have a component that uses polymorphism to render a Select component conditionally. It works by checking whether the runtime type is a subtype of the NodeModel. If so, the Select component is rendered, and if not, it is hidden. 
This part of the implementation works fine, but after having rendered the Select component and clicking any other form component (for example, the TextEdit form component), the runtime type resets to the supertype; hence the rendered Select component disappears. This seems to be caused by something in Blazorise or maybe Blazor itself (but i would assume using Polymorphism is supported by Blazor). 

```csharp
<TableRowCell>
    <Validation Validator="@(e => (e.Status, e.ErrorText) = FluentValidator.Validate(_validator, Node, nameof(Node.Name)))">
        <TextEdit Placeholder="Type a name..." TextChanged="UpdateName" Value="@Node.Name">
            <Feedback>
                <ValidationError/>
            </Feedback>
        </TextEdit>
    </Validation>
</TableRowCell>
<TableRowCell>
    <SelectType Node="@Node" SelectedValueChanged="@UpdateNode"/>
</TableRowCell>
<TableRowCell>
     @if (Node is ActorModel)
     {
          <Validation Validator="@(e => (e.Status, e.ErrorText) = FluentValidator.Validate(_validator, (ActorModel)Node, nameof(ActorModel.Role) + "." + nameof(ActorModel.Role.Id)))">
              <SelectRole Role="@(((ActorModel)Node).Role)" SelectedValueChanged="@UpdateRole"/>
          </Validation>
     }
</TableRowCell>
```

Example: 
In the code above, the NodeModel has two subtypes: the ObjectModel or the ActorModel. Setting any of these with my custom SelectType component, I want to be able to change the name, description, etc., without resetting the type back to NodeModel, which hides the SelectRole component.

**Comment by vbray1979** [comment]
> Can you provide an example?

OK here is the minimal code I have to reproduce the issue:
Notice that remove the reference to the validation model prevent the date resets.

Razor:
```
 <Validations @ref="_Validations" Mode="ValidationMode.Manual" Model="@NewContactModel">
            <Fields>
                <Validation>
                    <Field ColumnSize="ColumnSize.IsThird">
                        <FieldLabel>Birthday</FieldLabel>
                        <FieldBody>
                            <DateEdit TValue="DateTime?" @bind-Date="@NewContactModel.Birthday" />
                        </FieldBody>
                    </Field>
                </Validation>
            </Fields>
        </Validations>
@code{
    Validations _Validations;
    ContactModel NewContactModel = new ContactModel();
  }
```
Model:
```
 public class ContactModel : IInsertable, IUpdatable
    {
        [Column(Order = 4)]
        public DateTime? Birthday { get; set; }
}
```

**Comment by stsrki** [comment]
@niem94 You got to be careful with `@if` statements inside of your markup. The thing is that anything inside if it will not be part of the DOM until the statement is valid. Meaning it will not respond to any state changes.

**Comment by stsrki** [comment]
@vbray1979 I can reproduce it now. It seems this will be long search for real culprit 😅

**Comment by stsrki** [comment]
The problem is actually larger than I thought. Blazorise works perfectly, but the way how date input works in browser is the actual problem. See https://stackoverflow.com/a/57467732/833106

Basically, while you enter date, key by key, the event `onchange` is not triggered until you start entering year part. On that moment `onchange` will trigger and since the entered value is still not valid the actual value in Blazorise will reset to default.

There are two ways to deal with this:

- allow the use of global option `ChangeTextOnKeyPress` so that value is triggered by `oninput` event. This still leaves problem with `onchange` though.
- I can bypass `onchange` with the use of additional event `onblur`. This works good, but the problem is that the value will be changed only after the input loses focus.

**Comment by stsrki** [comment]
It pains me to say but it seems I will need to postpone fixing this ticket for some other time. I have being investigating it for the last couple of days and it's almost impossible to make native date input work nice. While entered values are perfectly fine for JS and html they are not OK for "business rules" and C#/Blazor.

Also I tried to see what other frameworks(DevExpress, Syncfusion, Telerik, MatBlazor) are doing, and it seems none of them are using `<input type="date">` but `<input type="text">` with bunch of JS behind.
